jax/core.py

```diff
@@ -261,6 +263,17 @@ def __complex__(self): return self.aval._complex(self)
   def __hex__(self): return self.aval._hex(self)
   def __oct__(self): return self.aval._oct(self)
 
+  # Like Python's checks for arithmetic special methods, NumPy doesn't check
+  # instance attributes when looking up __array_ufunc__ and __array_function__,
+  # so these need to be defined here rather than on UnshapedArray.
```

Comment:
Alternatively, if you really want to keep with dedicated methods on `DeviceArray` I think this could be solved by implementing `__getattr__` on a metaclass. I'm not sure you want to go down that dark path, though....

Reply:
It seems like the best we could do is forward to `self.aval`, but we'll still need the explicit `__array_ufunc__` and `__array_function__` function methods. It wouldn't do us any good to put these on a metadata, because then they would only have access to the `type` object (and `self.aval` is an instance variable).